src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -167,29 +137,40 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
   /**
    * Calculates campaign-wide statistic counters, also taking retry campaings
    * into account if any exist.
-   * TODO (OTA-2384) this method needs separate refactoring
    */
   def campaignStats(campaignId: CampaignId): Future[CampaignStats] = db.run {
+    final case class Counts(
+      processed: Long,
+      affected: Long,
+      rejected: Long,
+      cancelled: Long,
+      finished: Long)
+
+    def processCounts(counts: Map[DeviceStatus, Int]): Counts = Counts(
+      processed = counts.values.sum.toLong,
+      affected = counts.filterKeys(_ != DeviceStatus.rejected).values.sum.toLong,
+      rejected = counts.getOrElse(DeviceStatus.rejected, 0).toLong,
+      cancelled = counts.getOrElse(DeviceStatus.cancelled, 0).toLong,
+      finished =
+        counts.getOrElse(DeviceStatus.successful, 0).toLong +
+        counts.getOrElse(DeviceStatus.failed, 0).toLong
+    )
+
     val statsAction = for {
       mainCampaign <- campaignRepo.findAction(campaignId)
       retryCampaignIds <- campaignRepo.findRetryCampaignIdsOfAction(campaignId)
-      mainProcessed <- countProcessedAction(Set(campaignId))
-      mainAffected <- countAffectedAction(Set(campaignId))
-      retryProcessed <- countProcessedAction(retryCampaignIds)
-      retryAffected <- countAffectedAction(retryCampaignIds)
-      mainCancelled <- countCancelledAction(Set(campaignId))
-      retryCancelled <- countCancelledAction(retryCampaignIds)
-      mainFinished  <- countFinishedAction(Set(campaignId))
+      mainCnt <- deviceUpdateRepo.countByStatus(Set(campaignId)).map(processCounts)
+      retryCnt <- deviceUpdateRepo.countByStatus(retryCampaignIds).map(processCounts)
       // TODO (OTA-2307) replace with failed devices groups when implemented
       failed <- findFailedDevicesAction(campaignId)
     } yield CampaignStats(
       campaign = campaignId,
       status = mainCampaign.status,
-      finished = mainFinished - (retryProcessed - retryAffected + retryCancelled),
+      finished = mainCnt.finished  - (retryCnt.rejected + retryCnt.cancelled),
       failed = failed,
-      cancelled = mainCancelled + retryCancelled,
-      processed = mainProcessed,
-      affected  = mainAffected - (retryProcessed - retryAffected)
+      cancelled = mainCnt.cancelled + retryCnt.cancelled,
+      processed = mainCnt.processed,
+      affected  = mainCnt.affected - retryCnt.rejected,
```

Comment:
Why don't we send `succeeded` as well so we avoid all calculations in the FE?

Reply:
This change is not a part of the PR. It is solely about refactoring the code.

Reply:
Ah I see.